### PR TITLE
Feature/object fit util

### PR DESCRIPTION
**Pull request by firstchair**
After using this utility in various projects I thought it would be handy to have it added here. 
The util can imitate the CSS `object-fit` functionally or it do calculations to draw images on canvas element.

### CODE REVIEW

File: src/utils/objectFit/objectFit.mdx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  const childRef = useRef<HTMLDivElement>(null);
+
+  const onResize = useCallback(() => {
+    if (!parentRef.current || !childRef.current) {
+      return;
+    }
+
+    const objectFit = contain(parentRef.current.offsetWidth, parentRef.current.offsetHeight, 1, 1);
+
+    childRef.current.style.cssText += objectFit.cssText;
```
Comment (ThaNarie) [inline src/utils/objectFit/objectFit.mdx:92]: Even though this is just an example, it could happen that if the `parentRef` changes, it appends the new cssText to an existing element.

I was thinking the csstext would be used to assign to the JSX with the `style` tag, but unfortunately that receives an object, and not a text string.

I wonder if it would be useful to have an example like:

```tsx
const { x: left, y: top, width, height } = contain(parentRef.current.offsetWidth, parentRef.current.offsetHeight, 1, 1);


<div
  ref={childRef}
  style={{ position: 'relative', outline: '1px solid blue', left, top, width, height }}
/>
```

Or maybe have a cssObj returned with those 4 renamed style props, so it can be used in a similar way to cssText;

Benefit over JSX is that it sets the individual style props (left, top, etc), instead of working with a "dumb" string.

File: src/utils/objectFit/objectFit.test.ts
Comment (ThaNarie) [inline src/utils/objectFit/objectFit.test.ts:1]: You're having two tests, one for cover, one for contain.
Both tests have the exact same input, and the exact same output.
How can you be 100% sure, from this tests, that both versions work correctly? :)

File: src/utils/objectFit/objectFit.test.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        x: 0,
+        y: 0,
+        width: 100,
+        height: 100,
+        scale: 2,
+        cssText: 'left:0px;top:0px;width:100px;height:100px;',
+      });
+    });
+
+    it('throws an error for non-positive input arguments', () => {
```
Comment (ThaNarie) [inline src/utils/objectFit/objectFit.test.ts:16]: Technically, `0` is positive.
Maybe reword (and also in other places) to "greater than 0" ?

**Review by ThaNarie — CHANGES_REQUESTED** [review]

File: src/utils/objectFit/objectFit.mdx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+```ts
+function objectFit(fit: 'contain' | 'cover') {
+  return (
+    parentWidth: number,
+    parentHeight: number,
+    childWidth: number,
+    childHeight: number,
+  ): { x: number; y: number; width: number; height: number; scale: number; cssText: string } => {
+    if ([parentWidth, parentHeight, childWidth, childHeight].some((value) => value <= 0)) {
```
Comment (leroykorterink) [inline src/utils/objectFit/objectFit.mdx:27]: Using logical operators will be significantly faster compared to `Array.some`.

```ts
if (parentWidth <= 0 || parentHeight <= 0 || childWidth <= 0 || childHeight <= 0) {
  throw new Error(`All arguments should have a positive value`);
}
```
